US Medical Insurance costs project

Step one: import packages and data to work on.

In [22]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

data=pd.read_csv('insurance.csv')

Step two: what do we have inside ?

In [23]:
print(data.shape)
print(data.dtypes)
print(data.columns)

for val in data.columns:
    
    print(data[val].value_counts(dropna=False))
    
print(data.head(5))

(1338, 7)
age           int64
sex          object
bmi         float64
children      int64
smoker       object
region       object
charges     float64
dtype: object
Index(['age', 'sex', 'bmi', 'children', 'smoker', 'region', 'charges'], dtype='object')
18    69
19    68
51    29
45    29
46    29
47    29
48    29
50    29
52    29
20    29
26    28
54    28
53    28
25    28
24    28
49    28
23    28
22    28
21    28
27    28
28    28
31    27
29    27
30    27
41    27
43    27
44    27
40    27
42    27
57    26
34    26
33    26
32    26
56    26
55    26
59    25
58    25
39    25
38    25
35    25
36    25
37    25
63    23
60    23
61    23
62    23
64    22
Name: age, dtype: int64
male      676
female    662
Name: sex, dtype: int64
32.300    13
28.310     9
31.350     8
30.800     8
30.875     8
          ..
20.100     1
29.545     1
37.900     1
25.520     1
24.090     1
Name: bmi, Length: 548, dtype: int64
0    574
1    324
2    240
3    157
4     25
5     18
Name: children,

So we have a dataset of 1338 rows and 7 variables. Some variables are objects
, maybe we will need to change them in numerical (sex, region, smoker).

In [24]:
# We write functions to change data into numerical stuff

def sex(var):
    
    if var=='male':
        
        return 0
    
    elif var=='female':
        
        return 1
        
    
    
def smoke(var):
    
    if var=='no':
        
        return 0
    
    elif var=='yes':
        
        return 1
    
def region(var):
    
    if var=='southeast':
        
        return 0
    
    elif var=='southwest':
        
        return 1
    
    elif var=='northwest':
        
        return 2
    
    elif var=='northeast':
        
        return 3
    
    
    
# Now we change those variables:
print(data.head(20))

data['sex']=data['sex'].apply(lambda x: sex(x))
data['smoker']=data['smoker'].apply(lambda x: smoke(x))
data['region']=data['region'].apply(lambda x: region(x))

print(data.head(20))
print(data.dtypes)

    age     sex     bmi  children smoker     region      charges
0    19  female  27.900         0    yes  southwest  16884.92400
1    18    male  33.770         1     no  southeast   1725.55230
2    28    male  33.000         3     no  southeast   4449.46200
3    33    male  22.705         0     no  northwest  21984.47061
4    32    male  28.880         0     no  northwest   3866.85520
5    31  female  25.740         0     no  southeast   3756.62160
6    46  female  33.440         1     no  southeast   8240.58960
7    37  female  27.740         3     no  northwest   7281.50560
8    37    male  29.830         2     no  northeast   6406.41070
9    60  female  25.840         0     no  northwest  28923.13692
10   25    male  26.220         0     no  northeast   2721.32080
11   62  female  26.290         0    yes  southeast  27808.72510
12   23    male  34.400         0     no  southwest   1826.84300
13   56  female  39.820         0     no  southeast  11090.71780
14   27    male  42.130  

We could simplify some variable and round them (charges, bmi)

In [26]:
data['charges']=data['charges'].apply(lambda x: int(x))
data['bmi']= data['bmi'].apply(lambda x: int(x))

print(data.dtypes)

age         int64
sex         int64
bmi         int64
children    int64
smoker      int64
region      int64
charges     int64
dtype: object


Now we will try to make a model able to predict charges depends on the others variables

In [34]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor

target=data['charges']
dataset=data.drop(['charges'], axis=1)

X_train, X_test, y_train, y_test=train_test_split(dataset, target, random_state=243, test_size=0.2)

model=RandomForestRegressor()
model.fit(X_train, y_train)
print(model.score(X_test, y_test))

0.8604204579956021


By changing the random_state, the score range is between 82 and 88 %.